src/Skynet.Server/Network/DatabaseExtensions.cs

```diff
@@ -26,21 +28,28 @@ public static List<MessageDependency> ToDatabase(this List<Dependency> dependenc
             return result;
         }
 
-        public static ChannelMessage ToPacket(this Message message, long accountId)
+        public static ChannelMessage ToPacket(this Message message, PacketService packets, long accountId)
         {
-            var packet = new ChannelMessage
-            {
-                Id = message.PacketId,
-                PacketVersion = message.PacketVersion,
-                ChannelId = message.ChannelId,
-                SenderId = message.SenderId ?? 0,
-                MessageId = message.MessageId,
-                SkipCount = 0, // TODO: Implement flags and skip count
-                DispatchTime = DateTime.SpecifyKind(message.DispatchTime, DateTimeKind.Local),
-                MessageFlags = message.MessageFlags,
-                FileId = 0, // Files are not implemented yet
-                PacketContent = message.PacketContent
-            };
+            Packet prototype = packets.Packets[message.PacketId];
+            if (prototype == null) throw new ArgumentException($"Could not find packet {message.PacketId:x2}", nameof(message));
+
+            if (!(prototype is ChannelMessage protoMessage)) 
+                throw new ArgumentException($"Packet {message.PacketId:x2} is not a channel message", nameof(message));
+
+            if (!message.MessageFlags.AreValid(protoMessage.RequiredFlags, protoMessage.AllowedFlags))
+                throw new ArgumentException($"Message flags {message.MessageFlags} are invalid for packet {message.PacketId:x2}");
+
+            ChannelMessage packet = (ChannelMessage)prototype.Create();
+            packet.Id = message.PacketId;
```

Comment:
Why not keep the inline version?

Reply:
That's syntactically not possible as we don't use the constructor directly anymore.

Reply:
how tf have i not seen that